### ISSUE TITLE
AWS KMS plugin requires long-lived credentials

### ISSUE THREAD

**Issue by ddaviesbrackett-dwp** [kroxylicious/kroxylicious#1282]
**Is your feature request related to a problem? Please describe.**
The brand new AWS KMS plugin requires an access key and a secret key as configuration, provided in plaintext on disk. AWS disrecommends long-lived credentials, and many organisations outright forbid their use, because compromise to long-lived credentials is difficult to detect.

I understand this is at least partially because the decision was taken to avoid including the AWS SDK and instead do pure HTTP interaction with AWS KMS itself.

**Describe the solution you'd like**
I'd like to be able to specify an AWS IAM role that the Kroxylicious AWS KMS plugin can request short-lived, automatically-rotating credentials for. I'm far from an expert in IAM, so these details may be wrong, but if the proxy binary is running in EC2 then the credentials can be [fetched from the instance medatadata](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html#instance-metadata-security-credentials); if the proxy binary is running in Amazon EKS, [IAM roles can be attached to service accounts ](https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html) but I don't see in the AWS docs a reference to how those credentials can be retrieved outside of using the AWS SDK. 

**Describe alternatives you've considered**
* You could use an AWS SDK I guess, though as the KMS plugin itself notes in a code comment there's a considerable dependency weight associated with that. 
* You could keep requiring long-lived credentials.
* You could deploy an additional binary, or require an additional binary to be deployed, from which short-lived credentials could be requested (sort of like a kerberos refresh token command I guess?). 

**Additional context**
(this space intentionally left blank)

**Comment by k-wall** [comment]
> if the proxy binary is running in EC2 then the credentials can be [fetched from the instance medatadata](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html#instance-metadata-security-credentials);

I am understanding that AWS lets you create an IAM role and attach it to an EC2 instance during or after launch.  This role is accessible to applications running on the EC2 instance via a link local address.

From a **quick** look, it appears doable.   `AwsKmsService`'s authentication mechanism needs to become pluggable so that it is capable of specifying alternative authentication mechanism:

what we have today would be re-expressed like this:

```yaml
kms: AwsKmsService                                            
kmsConfig:
  endpointUrl: https://kms.<region>.amazonaws.com             
  authenticationMechanism: AccessKey
  accessKey:
    passwordFile: /opt/aws/accessKey                          
  secretKey:
    passwordFile: /opt/aws/secretKey                 
```

The use cases using an IAM Role, it would be something like this:

```yaml
kms: AwsKmsService                                            
kmsConfig:
  ...             
  authenticationMechanism: IamRole
  role: foo
```

The `AwsKmsService` would take responsibility to call the instance's link local address to get the token and get an accessKey/secretKey using it.  It would also take responsibility to refresh the credentials before they [expire](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html#:~:text=We%20make%20new%20credentials%20available%20at%20least%20five%20minutes%20before%20the%20expiration%20of%20the%20old%20credentials.) (background thread required).

Basically I think we'd need to code the equivalent of the [`curl` statements](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html).    On the surface, it doesn't look too hard.   We'd need to think about integration testing.  

I haven't looked at EKS integration yet.  I also want to understand how similar/different things are for [ROSA](https://www.redhat.com/en/technologies/cloud-computing/openshift/aws)



, https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html

**Comment by ddaviesbrackett-dwp** [comment]
selfishly, we're currently deploying on EC2 so that's the bit we're most interested in - and that we'd find easiest to prove :)

**Comment by k-wall** [comment]
Let's split this issue into two to track EC2 and EKS requirement separately.

**Comment by ddaviesbrackett-dwp** [comment]
#1294 and #1295 accurately represent the two issues here, closing.
